File: backend/src/Loupe.Application/ReferenceDrafts/ReferenceDraftResult.cs
using Loupe.Domain.References;
using Loupe.Application.Operations;

namespace Loupe.Application.ReferenceDrafts;

public sealed record ReferenceDraftResult(Guid Id, string Title, string? SourceUrl, string? Attribution,
    int? Width, int? Height, string? ImageUrl, string? PreviewUrl, long Revision, DateTimeOffset ExpiresAt, Guid? CommittedReferenceId, string? FailureCode, OperationResult? Import)
{
    public static ReferenceDraftResult From(ReferenceDraft draft) => new(draft.Id, draft.Title, draft.SourceUrl, draft.Attribution,
        draft.Width, draft.Height, draft.ImageKey is null ? null : $"/api/reference-drafts/{draft.Id}/image",
        draft.PreviewKey is null ? null : $"/api/reference-drafts/{draft.Id}/preview", draft.Revision, draft.ExpiresAt, draft.CommittedReferenceId,
        draft.FailureCode ?? draft.ImportOperation?.FailureCode, draft.ImportOperation is null ? null : OperationResult.From(draft.ImportOperation));
}

File: frontend/projects/api/src/lib/reference-draft/reference-draft-result.ts
import { OperationResult } from '../operation/operation-result';

export interface ReferenceDraftResult {
  id: string;
  title: string;
  sourceUrl: string | null;
  attribution: string | null;
  width: number | null;
  height: number | null;
  imageUrl: string | null;
  previewUrl: string | null;
  revision: number;
  expiresAt: string;
  committedReferenceId: string | null;
  failureCode: string | null;
  import: OperationResult | null;
}
